### DISCUSSION TITLE
zk-like fts feature

### DISCUSSION THREAD

**Discussion by gour** [getreu/tp-note#19] [Ideas]
Hello!

It looks that I'm somehow not compatible with Emacs' ecosystem and more inclined to use modal editors like (neo)vim, helix...That's the reason why I (again) explored Tp-Note. :smile: 

Although Denote does support signatures, it seems they do not consider them as important as you do: "The line of thinking, is one of the core principles of Luhmann's Zettelkasten." So, for me, that is one of the killing features of Tp-Note. :sunglasses: 

Now I also do like [zk](https://github.com/zk-org/zk) and it full-text-search (fts) capabilities stemming from its usage of sqlite3 database.

Tried to combine the two packages into single workflow, but it does not work, so I'm wondering if adding fts to Tp-Note could fit its scope?

Another feature I did already  mentioned is support for [[WikiLinks]] format which is practically becoming part of every single note-taking application (Logseq, Obsidian, Roam...). One of the best markdown LSP servers [Marksman](https://github.com/artempyanykh/marksman) also does support it as well as `zk` mentioned above...

However, I admit, it is possible to live **without** [[WikiLink]] support. :wink:

**Comment by getreu** [comment]
Hello!

nice to hear from you.

> Although Denote does support signatures, it seems they do not consider them as important as you do: "The line of thinking, is one of the core principles of Luhmann's Zettelkasten." So, for me, that is one of the killing features of Tp-Note. 😎
> 
> Now I also do like [zk](https://github.com/zk-org/zk) and it full-text-search (fts) capabilities stemming from its usage of sqlite3 database.

Tp-Note does not store it's notes in a database. Instead, they are just plain text UTF-8 files. Therefor, you can use your OS tool for full text search. Under Linux for example, you have `find` and most modern file manager have a similar find function.

> Tried to combine the two packages into single workflow, but it does not work, so I'm wondering if adding fts to Tp-Note could fit its scope?

Unlike `zk`, Tp-Note can only create, display and browse note files (following links in the note's markup rendition). The user can decide what programs he want to use for editing or retrieving notes according to his personal preferences and taste. This is why the best place to add full text search with a plugin, is either your file manager, or your file editor. For example, I am using the internal search function of the file manager [Thunar](https://wiki.archlinux.org/title/Thunar#Search_for_files_and_folders). The Helix file editor for example has an excellent fuzzy matcher called [nucleo](https://github.com/helix-editor/nucleo). More search options and programs are described [here](https://blog.getreu.net/20230904-tp-note-news11/#retrieving-notes).

> Another feature I did already mentioned is support for [[WikiLinks]] format which is practically becoming part of every single note-taking application (Logseq, Obsidian, Roam...). One of the best markdown LSP servers [Marksman](https://github.com/artempyanykh/marksman) also does support it as well as `zk` mentioned above...
> 
> However, I admit, it is possible to live **without** [[WikiLink]] support. 😉

For far, the Wikilink syntax has not found it's way into the CommonMark standard and is not  (yet) supported. Markdown has it's own 
shorthand link syntax: `<http:/mypage.md>`.

Tp-Note has a dedicated support for Zettelkasten - hyperlinks, e.g. `<tpnote:5b?>` which is displayed as _My page_. The example link points to another document with the ID (sort-tag)  `5b` and the title _My page_. More examples in [Linking notes and resources](https://blog.getreu.net/20230904-tp-note-news11/#linking-notes-and-resources).

**Comment by gour** [comment]
> nice to hear from you.

:smiley: 

> Tp-Note does not store it's notes in a database. Instead, they are just plain text UTF-8 files. Therefor, you can use your OS tool for full text search. Under Linux for example, you have `find` and most modern file manager have a similar find function.

Ahh, correct. I was completely enamoured with sqlite's fts feature. :disappointed: 

> This is why the best place to add full text search with a plugin, is either your file manager, or your file editor. For example, I am using the internal search function of the file manager [Thunar](https://wiki.archlinux.org/title/Thunar#Search_for_files_and_folders). 

Currently, I'm on GNOME due to persuasion from some of my family members, but could switch back to Xfce at any time...

> The Helix file editor for example has an excellent fuzzy matcher called [nucleo](https://github.com/helix-editor/nucleo). More search options and programs are described [here](https://blog.getreu.net/20230904-tp-note-news11/#retrieving-notes).

Thank you for the reference! I went several times of the article, but still missing important points...

> For far, the Wikilink syntax has not found it's way into the CommonMark standard and is not (yet) supported. Markdown has it's own shorthand link syntax: `<http:/mypage.md>`.
> 
> Tp-Note has a dedicated support for Zettelkasten - hyperlinks, e.g. `<tpnote:5b?>` which is displayed as _My page_. The example link points to another document with the ID (sort-tag) `5b` and the title _My page_. More examples in [Linking notes and resources](https://blog.getreu.net/20230904-tp-note-news11/#linking-notes-and-resources).

I really **like** that you're taking Luhman's naming scheme seriously.

Have you seen [neorg](https://github.com/nvim-neorg/neorg) project - pretty ambitious creating a new markup format on the org-mode' foundation, but just curious how hard it would be to add support for it to Tp-Note?

**Comment by getreu** [comment]
> Currently, I'm on GNOME due to persuasion from some of my family members, but could switch back to Xfce at any time...

You do not need to leave GNOME for Thunar. As Thunar is based on the Gnome libraries, you can install it in addition without pulling extra other libraries. Anyway, any file manager will do. Examples [here](https://blog.getreu.net/projects/tp-note/tpnote--manual.html#linux-file-manager-configuration).

> Have you seen [neorg](https://github.com/nvim-neorg/neorg) project - pretty ambitious creating a new markup format on the org-mode' foundation, but just curious how hard it would be to add support for it to Tp-Note?

I had a look. The markup language looks very similar to reStructuredText (`.rst`) to me. Is it so? Tp-Note has some [basic support for .rst note files](https://blog.getreu.net/projects/tp-note/tpnote--manpage.html#change-the-default-markup-language-to-restructuredtext).
